### DISCUSSION TITLE
Does Kombu support channel guarantee mode of confirm?

### DISCUSSION THREAD

**Discussion by jackson-chris** [celery/kombu#1995] [Q&A]
From what I've read in documentation if using `pyamqp` transport I should be able to set the `confirm_publish` arg to `True` to ensure that any channels that get opened with that connection will use the channel guarantee mode `Confirm`. 

I tried this against a local RMQ cluster I had by running the following block of code:
```
from kombu import Connection
import os

RABBITMQ_USERNAME = os.getenv("RABBITMQ_USERNAME", "guest")
RABBITMQ_PASSWORD = os.getenv("RABBITMQ_PASSWORD", "guest")
RABBITMQ_HOST = os.getenv("RABBITMQ_HOST", "localhost")
RABBITMQ_PORT = os.getenv("RABBITMQ_PORT", "5672")
RABBITMQ_SCHEME = os.getenv("RABBITMQ_SCHEME", "amqp")
RABBITMQ_VHOST = 'test'
RABBITMQ_URL = (
    os.getenv("RABBITMQ_URL")
    or f"{RABBITMQ_SCHEME}://{RABBITMQ_USERNAME}:{RABBITMQ_PASSWORD}@{RABBITMQ_HOST}:{RABBITMQ_PORT}/{RABBITMQ_VHOST}"
)

connection = Connection(RABBITMQ_URL, confirm_publish=True, transport_options={'confirm_publish':True}, transport='pyamqp')
connection.connect()
channel = connection.channel()
```

I then looked at the RMQ management UX at a list of connections/channels for that particular vhost and see that a connection and channel were created. However the channel does NOT show the correct mode of `Confirm`. Am I missing something in my connection declaration? 

Details about packages used for testing:
```
# pip show kombu
Name: kombu
Version: 5.3.7
Summary: Messaging library for Python.
Home-page: https://kombu.readthedocs.io
Author: Ask Solem
Author-email: auvipy@gmail.com
License: BSD-3-Clause
Location: /usr/local/lib/python3.9/site-packages
Requires: typing-extensions, vine, amqp
Required-by:

# pip show amqp
Name: amqp
Version: 5.2.0
Summary: Low-level AMQP client for Python (fork of amqplib).
Home-page: http://github.com/celery/py-amqp
Author: Barry Pederson
Author-email: auvipy@gmail.com
License: BSD
Location: /usr/local/lib/python3.9/site-packages
Requires: vine
Required-by: kombu
```

I saw this [issue](https://github.com/celery/kombu/issues/572) which seemed relevant.

**Comment by jackson-chris** [comment]
It would seem I need to call:
```
channel.confirm_select()
```
Once doing so I properly see the correct mode set for the channel.

My follow up question is what is the best way to ensure this mode is used for all channels associated with a https://docs.celeryq.dev/projects/kombu/en/stable/reference/kombu.mixins.html#kombu.mixins.ConsumerProducerMixin class?

**Comment by endotronic** [comment]
I have the exact same question as you. I'm looking to achieve at-least-once processing with RMQ and I think the Confirm mode is necessary for this. I'm not finding this `confirm_select` function you mention though. I'm naively looking here: https://github.com/celery/kombu/blob/dcb43be8916865e066643a3b8fd1bd6fb767601a/kombu/transport/pyamqp.py#L102

How/where did you find this function? Personally I am not using the mixin, so this solution would possibly be enough for me, though I would also want some guarantee that `publish` will block until the confirm and raise if it isn't confirmed.

FWIW you may have an easy time moving away from the mixin. See this writeup (not mine and kudos to the author): https://medium.com/python-pandemonium/building-robust-rabbitmq-consumers-with-python-and-kombu-part-1-ccd660d17271
and the associated code: https://github.com/Skablam/kombu-examples/tree/master
